packages/xstate-solid/test/useService-fsm.test.tsx

```diff
@@ -0,0 +1,174 @@
+      }
+    }
+  });
+
+  it('should share a single service instance', () => {
+    const counterService = interpret(counterMachine()).start();
+
+    const Counter = () => {
+      const [state] = useService(() => counterService);
+
+      return <div data-testid="count">{state.context.count}</div>;
+    };
+
+    render(() => (
+      <div>
+        <Counter />
+        <Counter />
+      </div>
+    ));
+
+    const countEls = screen.getAllByTestId('count');
+
+    expect(countEls.length).toBe(2);
+
+    countEls.forEach((countEl) => {
+      expect(countEl.textContent).toBe('0');
+    });
+
+    counterService.send('INC');
+
+    countEls.forEach((countEl) => {
+      expect(countEl.textContent).toBe('1');
+    });
+  });
+
+  it('service should be updated when it changes', () => {
+    const counterService1 = interpret(counterMachine()).start();
+    const counterService2 = interpret(counterMachine()).start();
+
+    const Counter = (props) => {
+      const [state, send] = useService(
+        props.counterRef
+      );
+
+      return (
+        <div>
+          <button data-testid="inc" onclick={(_) => send('INC')} />
+          <div data-testid="count">{state.context.count}</div>
+        </div>
+      );
+    };
+    const CounterParent = () => {
+      const [service, setService] = createSignal(counterService1);
+
+      return (
+        <div>
+          <button
+            data-testid="change-service"
+            onclick={() => setService(() => counterService2)}
+          />
+          <Counter counterRef={service} />
+        </div>
+      );
+    };
+
+    render(() => <CounterParent />);
+
+    const changeServiceButton = screen.getByTestId('change-service');
+    const incButton = screen.getByTestId('inc');
+    const countEl = screen.getByTestId('count');
+
+    expect(countEl.textContent).toBe('0');
+    fireEvent.click(incButton);
+    expect(countEl.textContent).toBe('1');
+    fireEvent.click(changeServiceButton);
+    expect(countEl.textContent).toBe('0');
+  });
+
+  it('service should be able to be used from useMachine', () => {
+    const CounterDisplay: Component<{
+      service: StateMachine.Service<any, any>;
+    }> = (props) => {
+      const [state] = useService(() => props.service);
+
+      return <div data-testid="count">{state.context.count}</div>;
+    };
+
+    const Counter = () => {
+      const [, send, service] = useMachine(counterMachine());
+
+      return (
+        <div>
+          <button data-testid="inc" onclick={(_) => send('INC')} />
+          <CounterDisplay service={service} />
+        </div>
+      );
+    };
+
+    render(() => <Counter />);
+
+    const incButton = screen.getByTestId('inc');
+    const countEl = screen.getByTestId('count');
+
+    expect(countEl.textContent).toBe('0');
+    fireEvent.click(incButton);
+    expect(countEl.textContent).toBe('1');
+  });
+
+  it('service should warn when reusing the same machine instance - reusing will result in shared context', () => {
```

Comment:
we don't assert this expected result anyhow in this test. I would do this:

1. override `console.warn` with `jest.mock()` at the beginning of this test
2. add `afterEach(() => jest.clearMocks())` at the top of the file
3. add smth like `expect(console.warn.mock.calls).toMatchInlineSnapshot()` at the end of this test

 I would we should call smth like `expect(console.warn).`

Reply:
We are calling `expect(warn).toBeCalled();` after the click events for this test. Would this change work for you? 

```typescript
expect(console.warn).toHaveBeenCalledWith(expect.stringContaining("...string with exact warning message"))
```

Reply:
ah, i've missed this. I would probably still move it to the end  - I assume that the test case is synchronous~, or at least synchronous-like. By putting it at the end we ensure that we actually run this assertion - assertions at different places (like in event handlers) have always a risk associated with them - the handler might not have been called for some reason.

I would also still prefer applying the 1st and 2nd suggestion. The 3rd one is debatable, so I don't mind whatever you choose here.